Add coverage for Quarkus #43422

jedla97 opened the pull request:
### Summary

Adding coverage for https://github.com/quarkusio/quarkus/issues/43422 issue.

The PR to fix the original issue is https://github.com/quarkusio/quarkus/pull/43440
 
Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Backport
- [x] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [x] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

jedla97 commented:
run tests

jedla97 commented:
Native is related will look at it

jcarranzan commented on http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GreetingResource.java:
```diff
@@ -1,8 +1,18 @@
+
+    @Path("/sub-resource")
+    public GreetingSubResources helloFromSubResource() {
+        return CDI.current()
```
Why use CDI directly instead of creating a field @Inject annotation? Are there some benefits to doing it here?

jcarranzan replied:
Ok I see that the issue reproducer use it, so I guess could be related to the issue and you had that motivation to use it.

jcarranzan reviewed (approved):
LGTM thanks

rsvoboda reviewed (changes requested):
* `OPTIONS` is not covered
 * what about other HTTP methods and sub-resource use case?
 * test failures need investigation

jedla97 commented:
Native issue was caused by this https://github.com/quarkusio/quarkus/issues/42976 so I create different resource class where I tested it + testing it with interface. 

Added `OPTION` request, what I tried it worked same even before the upstream PR.
Added `POST` method to sub-resource to check if it works (Personaly think that using it like this is more like edgecase)

jedla97 commented:
run tests

jcarranzan reviewed (commented):
Thanks @jedla97 the tests look good to me and now are covering all topics mentioned.
I have two small suggestions:
1. Could you consider creating a reusable helper method since most of tests have similar structure body?
2. On the other hand, I am wondering if it would be good having some negative test that returns for instance a 405 status code where be appropiated.

jedla97 commented:
> Could you consider creating a reusable helper method since most of tests have similar structure body?

@jcarranzan I created the methods for HEAD and OPTION requests. With that I combined them into one test.

> On the other hand, I am wondering if it would be good having some negative test that returns for instance a 405 status code where be appropiated.

Tbh I don't see much added value in these negative tests. Also what should be testing? Not able to send `DELETE` request as it's not defined on endpoint? If you insist on adding something like this I can but as I said I don't see any adding value as I believe that this is tested by unit test in Quarkus/Vert.x/RestEasy

jedla97 commented:
run tests

jcarranzan commented:
Ok thanks, let's wait for the checks.

michalvavrik commented on http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java:
```diff
@@ -102,6 +104,64 @@ public void abstractResourceWithPath() {
+    @Test
+    @Tag("https://github.com/quarkusio/quarkus/issues/43422")
+    @DisplayName("Test Quarkus REST using CDI getting the sub-resource for OPTION and HEAD request")
+    void cdiSubResourceOptionAndHeadRequest() {
```
This is only method that fails with 3.15.1. Why are other tests `optionAndHeadRequestUsingInterfaceResourceWithPath` and `optionAndHeadRequestUsingAbstractResourceWithPath` tagged with `https://github.com/quarkusio/quarkus/issues/43422` if they are passing with 3.15.1?

jedla97 replied:
I added the tag there as this issue was main motivation to create that coverage. If you things tha tag shouldn't be there. I can create PR for its removal.

michalvavrik replied:
Clear answer :-D : 

I don't really care, keep it as it. I just thought that if you tag something with an issue, it means that issue is tested by this. It can still be true that changed code is not failing but testing changes because it can test code that we didn't have covered previously. What I do when I verify bug tickets is that I look for tests tagged with that issue. Now I'll keep in mind that failing test tagged with an issue may not mean that issue is not fixed. It probably won't happen anyway.

michalvavrik replied:
Also, why are other 3 tests added in this PR without this tag, is that because they were not added because of this issue or you just forgot?

rsvoboda replied:
Goodcatch @michalvavrik. @jedla97 can you prepare a followup PR with backport label?

jedla97 replied:
> Also, why are other 3 tests added in this PR without this tag, is that because they were not added because of this issue or you just forgot?

There are more like to ensure that endpoint work. The quarkus fix was to add `HEAD` and `OPTION`. At least this is my logic behind adding the tag here.

> @jedla97 can you prepare a followup PR with backport label?

@rsvoboda with adding or removing the tags?

michalvavrik replied:
I think this kinda depends on how other QE members use `@Tag`, I always thought that it signals that some test tests the issue/ticket. But maybe that was just me.

rsvoboda replied:
`@Tag` usage ... there is no rule really, some members started to use it, but we didn't have any discussion and  conclusion on its usage.

I would personally use `@Tag` only on tests testing the bug. I wouldn't tag the other related tests.

michalvavrik commented:
@jedla97 I think we should try to have PR titled with a text that describes what PR does, you checked https://cbea.ms/git-commit/, `coverage for Quarkus #43422` is less informative then `Test support for @HEAD and @OPTIONS in sub-resources` or something like that. I realize sometimes it is really hard to write correct title (and I often fail).
